### PR TITLE
CMake subprojects

### PR DESCRIPTION
**Pull request by mensinda**
This PR adds support for using CMake projects as subprojects (without requiring a `meson.build`).

## Basic principle

 1. Configure the subproject with CMake
 2. Start the CMake server
 3. Extract the codemodel from the server
 4. Post process the codemodel
 5. Generate a meson AST
 6. Use the existing submodule architecture to execute the AST

## Current limitations:

 - Projects that rely on `add_custom_command()` and/or other code generators won't work
 - Everything else I have missed and don't know (yet).

## Tested projects:

This PR was tested with the following projects:

 - SDL (2.0.9) with `cmake_options: ['-DVIDEO_WAYLAND=OFF']` (wayland uses a code generator)
 - glslang (master) doesn't work because of a single `.y` file meson doesn't recognize

If someone has tested other projects with this PR, please let me know.

### CODE REVIEW

File: mesonbuild/cmake/interpreter.py [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+from .. import mlog
+from ..build import Build
+from ..environment import Environment
+from ..mparser import Token, BaseNode, CodeBlockNode, FunctionNode, ArrayNode, ArgumentNode, AssignmentNode, BooleanNode, StringNode, IdNode, MethodNode
+from ..backend.backends import Backend
+from ..compilers.compilers import obj_suffixes
+from ..dependencies.base import CMakeDependency, ExternalProgram
+from subprocess import Popen, PIPE, STDOUT
+from typing import List
+import os, re, shutil
```
Comment (ignatenkobrain) [inline mesonbuild/cmake/interpreter.py]: __[Flake8]__ [F401] 'shutil' imported but unused

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
I haven't looked too closely at the implementation, but as we discussed on IRC and from your proposal this seems like a fine approach, and probably very useful. The one thing I think I'd prefer is to have a module to handle this rather than being part of subproject.

My reasoning is that a subproject generally has a in meson defined dependency:
```meson
dep = declare_dependency(link_with : lib, include_directories : incs)
```

CMake obviously doesn't have that in the form we would expect, Since we may not want to build all of the targets in the cmake build definition, a module would make it easier to take the AST, store it, and then pull out only the bits that we want so if we wanted two targets out of a CMake build (but at different times or under different condtions) we could do that.

Ultimately this is a @jpakkane decision.

P.S. This is really interesting to me, and if we land this I think I'll work on a version for make/automake that does something similar.

**Comment by xclaesse** [comment]
I haven't looked at the patches themself, but my PR also allows building cmake subprojects: https://github.com/mesonbuild/meson/pull/4321

**Comment by dcbaker** [comment]
@xclaesse, I think this is a better approach, as the goal is to use cmake to configure the project, then use cmake's introspection API to create a meson representation and allow the subproject to be built as part of the main ninja call like a native meson subproject. If it can be made to work reliably of course :)

**Comment by mensinda** [comment]
@xclaesse I haven't looked at #4321 before, but the PR's seem to have two different approaches (as @dcbaker has already pointed out).

I am currently a bit short on time (I have two more exams), so I am not sure when I have time again to continue working on this PR.

**Comment by mensinda** [comment]
This PR now also works on Windows so it is no longer WIP.

There are some weird issues with the msvc2015 CI that I can't resolve so I disabled the tests there.

**Comment by mensinda** [comment]
ping :)

**Comment by ignatenkobrain** [comment]
Flake8 detected 6 issues on 926269e3fdce943939bd8b31f7e11d8375aab3a2
Visit https://sider.review/gh/repos/19784232/pulls/4969 to review the issues.

<sub>posted by Sider</sub>

**Comment by FelixSchwarz** [comment]
Just a question from a meson newbie: How do include dirs work with this patch?
I built a simple exe which links to docopt.cpp (cmake) as a subproject but compilation of my executable fails because there is no "-I" pointing to docopt? My workaround was to use `executable(..., subprojects/docopt.cpp)`

Other than that your code worked great for me - even cross-compiling with Windows worked. :-)

(Excuse me if this comment derails the pull request. If you want me to post in a different - or not at all - please just tell me.)

**Comment by mensinda** [comment]
@FelixSchwarz for each CMake target, this PR generates a meson target with the same name and the variables `targetname_inc` for the include directories, `targetname_src` for the source list (though this might not be useful) and `targetname_dep` which should be used as a dependency in your `meson.build`.

Basically this CMake code:
```cmake
set(source_list "docopt.cpp")
include_directories(...)
add_executable(foo ${source_list})
```
is converted into this meson code:
```meson
foo_inc = include_directories('__CMake_build', '.', ...)
foo_src = files('docopt.cpp')
foo = executable(
  'foo',
  foo_src,
  link_args : [],
  link_with : [],
  include_directories : foo_inc,
  install : false,
  install_dir : '',
  override_options : [],
  objects : [],
  cpp_args : []
)
foo_dep = declare_dependency(
  link_args : [],
  link_with : foo,
  include_directories : foo_inc
)
```

Then, something like 
```meson
executable('mainPro', main_src, dependencies: [subPro.get_variable('foo_dep')])
```
in your main project should work 99% of the time.

You can also view the generated meson code for the subproject in `buildDir/meson-logs/meson-log.txt`.

Could you please provide an example repository, if this didn't answer your question?

**Comment by FelixSchwarz** [comment]
@mensinda That was very helpful. It works for me now :-)

**Comment by mensinda** [comment]
ping

**Comment by jpakkane** [comment]
There is one major reason why I don't like this, and it is mentioned in the PR itself. There is no way to make this reliable. It does work for simple cases but if we make it a top level construct, then people will (quite reasonably) expect it to work a lot better than it probably can ever be made to work. This will bring with it a massive maintenance burden because we will then be getting a steady stream of bug reports about CMake subprojects that don't work.

Because of this we probably can't put this directly in `subproject`.

However maybe there is a way we could provide for this but in a way that makes it more obvious that this is not a first class supported case but instead something that you can try and which works on a "some effort", "if it breaks you get to keep both pieces" sort of way. Such as moving the invocation from `subproject` et al to the `cmake` module and give it a more appropriate name something in the vein of `best_effort_attempt_to_build_a_cmake_project`?

My personal opinion is and remains that trying to mix two build systems in one build directory is a fool's errand and won't work. Something like Flatpak-builder's staged manifest method is a superiour and "correct" solution to this problem. Sadly Windows and macOS will never provide for something similar so we have to keep supporting crappy solutions, probably until the end of time. :(

**Comment by dcbaker** [comment]
I would also prefer to see this in a module in part because of @jpakkane's concerns, and in part because supporting cmake isn't a core goal of meson. If we can make this work for a reasonable subset of projects then I think it would be useful to have, but I've seen a lot of batshit crazy cmake from proprietary land (sh/bat invokes cmake invokes sh/bat invokes cmake), and even from open source land (the llvm cmake build generates cmake files and sources them during the build process). So, we'll never be able to support every single cmake build.

**Comment by mensinda** [comment]
Sure, I can do that. This also has the potential to improve usability/readability.

Would it be OK if I add one new method in the already existing CMake module as well as a new returned object? I am thinking of something like this:

```meson
cmake = import('cmake')
sub = cmake.subproject('foo')

exe1 = executable('exe1', ['sources'], dependencies: sub.dependency('foo_target'))
```

**Comment by mensinda** [comment]
I have moved the subprojects interface into the CMake module. However, I mostly kept the actual subprojects code in the interpreter to avoid code duplication.

File: docs/markdown/CMake-module.md [hunk 2]
```diff
@@ -10,6 +12,91 @@ following functions will then be available as methods on the object
...
+
+# Fetch the dependency object
+cm_lib = sub_proj.dependency('cm_lib')
+
+executable(exe1, ['sources'], dependencies: [cm_lib])
+```
+
+The `subproject` method is almost identical to the normal meson
+`subproject` function. The only difference is that a CMake project
+instead of a meson prokect is configured.
```
Comment (TheQwertiest) [inline docs/markdown/CMake-module.md]: s/"prokect"/"project"

File: docs/markdown/CMake-module.md [hunk 3]
```diff
@@ -10,6 +12,91 @@ following functions will then be available as methods on the object
...
+)
+```
+
+It should be noted that not all projects are guaranteed to work. The
+safest approach would still be to create a `meson.build` for the
+subprojects in question.
+
+### `subproject` object
+
+This is object is returned by the `subproject` function described above
```
Comment (TheQwertiest) [inline docs/markdown/CMake-module.md]: s/"This is object is"/"This object is"

File: docs/markdown/CMake-module.md [hunk 4]
```diff
@@ -10,6 +12,91 @@ following functions will then be available as methods on the object
...
+and supports the following methods:
+
+ - `dependency(target)` returns a dependency object for any CMake target.
+ - `include_directories(target)` returns a meson include_directories
```
Comment (TheQwertiest) [inline docs/markdown/CMake-module.md]: maybe s/"include_directories"/"\`include_directories()\`"?

**Comment by eraxillan** [comment]
Ping!
@mensinda @TheQwertiest Any plans to merge this awesome feature to upstream?
Such ability to link cmake projects will allow many people to switch to meson, i think

**Comment by mensinda** [comment]
@eraxillan Yes, this PR is planned to go upstream. However, the review process takes some time since this is a relatively large PR (ping @jpakkane).

I have also created a small [meson/cmake project](https://github.com/mensinda/cmake_meson_test) for this PR to demonstrate some edge cases that are currently supported. It uses my old [CMake library](https://github.com/mensinda/cmakeBuildTools) that pretty much does all the things meson doesn't stand for (like automatically detecting source files and "parsing" C++ code with the CMake script language). @dcbaker This project also generates CMake files and sources them, so one of your crazy CMake use cases is already supported 😄.

The only real thing that currently doesn't work is the `add_custom_command()` related stuff. However, I already have an idea on how to get this working in a followup PR.

@TheQwertiest All typos should be fixed now. Thanks.

CI errors seem to be unrelated.

File: mesonbuild/interpreter.py [hunk 5]
```diff
@@ -2020,7 +2021,7 @@ def get_cross_property_method(self, args, kwargs):
                     'both_libraries': known_library_kwargs,
                     'library': known_library_kwargs,
                     'subdir': {'if_found'},
-                    'subproject': {'version', 'default_options', 'required'},
+                    'subproject': {'version', 'default_options', 'required', 'method', 'cmake_options'},
```
Comment (jpakkane) [inline mesonbuild/interpreter.py]: Since the functionality was moved to the module, these should no longer be here, right?

Comment (mensinda) [reply mesonbuild/interpreter.py]: Right. Fixed with the last commit.

**Comment by jpakkane** [comment]
The Cygwin failure seems to be relevant:

```
assertion "0" failed: file "/cygdrive/d/cyg_pub/devel/libuv/libuv-1.24.1-1.x86_64/src/libuv-1.24.1/src/unix/core.c", line 182, function: uv_close
assertion "0" failed: file "/cygdrive/d/cyg_pub/devel/libuv/libuv-1.24.1-1.x86_64/src/libuv-1.24.1/src/unix/core.c", line 182, function: uv_close
Stack trace:
Frame        Function    Args
000FFFFB0B0  00180060BF9 (00000000000, 6A83B3641190, 00000000001, 000FFFFDE50)
00000000000  00180062AD7 (00000000064, 00000000000, 00000000000, 00000000000)
000FFFFB3A0  001801314DB (00000000001, 00000000060, 00000000000, 00000000000)
000FFFFB720  001801575E4 (00000000000, 00000000000, 00000000000, 00000000000)
000FFFFB720  0018012DB7B (7FF85CEF5EC2, 00000004048, 00180249508, 0018013374C)
000FFFFB720  0018012DEAF (003F3D99174, 003F3D990B8, 000000000B6, 0018024948F)
000FFFFB720  001800433E5 (00000000000, 00000000000, 7FF85CF217AA, 006000F6000)
000FFFFB860  003F3D87459 (000FFFFB7F0, 0018030BBD0, 000FFFFB870, 006000DF540)
000FFFFB860  00100707C50 (00180142A05, 00000000003, 00100000001, 006000DF510)
000FFFFB860  00100707C2E (00600058A30, 000FFFFB990, 000FFFFB900, 000FFFFC330)
000FFFFB860  001006C025A (000FFFFB920, 00000000000, 00000000000, 00180142A8F)
000FFFFB860  001006B75FA (000FFFFB870, 00000000001, 000FFFFB900, 000FFFFB860)
000FFFFB860  003F3D86EC1 (00000000002, 001FFFFFFFF, 000FFFFBC80, 00600058D50)
00000000000  003F3D951FF (00180154381, 000FFFFC338, 00180157335, 000FFFFBD70)
00000000000  003F3D87648 (000FFFFC330, 00000000000, 006000589B8, 00000000001)
000FFFFC380  001006B40D2 (000FFFFBF88, 000000004B4, 000000517AC, 000FFFFBEE1)
End of stack trace (more stack frames may be present)
assertion "0" failed: file "/cygdrive/d/cyg_pub/devel/libuv/libuv-1.24.1-1.x86_64/src/libuv-1.24.1/src/unix/core.c", line 182, function: uv_close
Stack trace:
Frame        Function    Args
000FFFFB0B0  00180060BF9 (00000000000, 75E6BF035E9F, 00000000001, 000FFFFDE50)
00000000000  00180062AD7 (00000000064, 00000000000, 00000000000, 00000000000)
000FFFFB3A0  001801314DB (00000000001, 00000000060, 00000000000, 00000000000)
000FFFFB720  001801575E4 (00000000000, 00000000000, 00000000000, 00000000000)
000FFFFB720  0018012DB7B (7FF85CEF5EC2, 00000004048, 00180249508, 0018013374C)
000FFFFB720  0018012DEAF (003F3D99174, 003F3D990B8, 000000000B6, 0018024948F)
000FFFFB720  001800433E5 (00000000000, 00000000000, 7FF85CF217AA, 006000F5A80)
000FFFFB860  003F3D87459 (000FFFFB7F0, 0018030BBD0, 000FFFFB870, 006000D7F80)
000FFFFB860  00100707C50 (00180142A05, 00000000003, 00100000001, 006000D7F50)
000FFFFB860  00100707C2E (00600058A30, 000FFFFB990, 000FFFFB900, 000FFFFC330)
000FFFFB860  001006C025A (000FFFFB920, 00000000000, 00000000000, 00180142A8F)
000FFFFB860  001006B75FA (000FFFFB870, 00000000001, 000FFFFB900, 000FFFFB860)
000FFFFB860  003F3D86EC1 (00000000002, 001FFFFFFFF, 000FFFFBC80, 00600058D50)
00000000000  003F3D951FF (00180154381, 000FFFFC338, 00180157335, 000FFFFBD70)
00000000000  003F3D87648 (000FFFFC330, 00000000000, 006000589B8, 00000000001)
000FFFFC380  001006B40D2 (000FFFFBF88, 000000004B4, 000000517AC, 000FFFFBEE1)
End of stack trace (more stack frames may be present)
assertion "0" failed: file "/cygdrive/d/cyg_pub/devel/libuv/libuv-1.24.1-1.x86_64/src/libuv-1.24.1/src/unix/core.c", line 182, function: uv_close
Stack trace:
Frame        Function    Args
000FFFFB0B0  00180060BF9 (00000000000, 958B503F0A3A, 00000000001, 000FFFFDE50)
00000000000  00180062AD7 (00000000064, 00000000000, 00000000000, 00000000000)
000FFFFB3A0  001801314DB (00000000001, 00000000060, 00000000000, 00000000000)
000FFFFB720  001801575E4 (00000000000, 00000000000, 00000000000, 00000000000)
000FFFFB720  0018012DB7B (7FF85CEF5EC2, 00000004048, 00180249508, 0018013374C)
000FFFFB720  0018012DEAF (003F3D99174, 003F3D990B8, 000000000B6, 0018024948F)
000FFFFB720  001800433E5 (00000000000, 00000000000, 7FF85CF217AA, 0060019E830)
000FFFFB860  003F3D87459 (000FFFFB7F0, 0018030BBD0, 000FFFFB870, 006001AB2C0)
000FFFFB860  00100707C50 (00180142A05, 00000000003, 00100000001, 006001AB290)
000FFFFB860  00100707C2E (00600058A30, 000FFFFB990, 000FFFFB900, 000FFFFC330)
000FFFFB860  001006C025A (000FFFFB920, 00000000000, 00000000000, 00180142A8F)
000FFFFB860  001006B75FA (000FFFFB870, 00000000001, 000FFFFB900, 000FFFFB860)
000FFFFB860  003F3D86EC1 (00000000002, 001FFFFFFFF, 000FFFFBC80, 00600058D50)
00000000000  003F3D951FF (00180154381, 000FFFFC338, 00180157335, 000FFFFBD70)
00000000000  003F3D87648 (000FFFFC330, 00000000000, 006000589B8, 00000000001)
000FFFFC380  001006B40D2 (000FFFFBF88, 000000004B4, 000000517AC, 000FFFFBEE1)
End of stack trace (more stack frames may be present)
Succeeded test: test cases/cmake/1 basic
Succeeded test: test cases/cmake/2 advanced
Succeeded test: test cases/cmake/3 advanced no dep
Succeeded test: test cases/cmake/4 code gen
assertion "0" failed: file "/cygdrive/d/cyg_pub/devel/libuv/libuv-1.24.1-1.x86_64/src/libuv-1.24.1/src/unix/core.c", line 182, function: uv_close
assertion "0" failed: file "/cygdrive/d/cyg_pub/devel/libuv/libuv-1.24.1-1.x86_64/src/libuv-1.24.1/src/unix/core.c", line 182, function: uv_close
Stack trace:
Frame        Function    Args
000FFFFB0B0  00180060BF9 (00000000000, 662AE5DD3750, 00000000001, 000FFFFDE50)
00000000000  00180062AD7 (00000000064, 00000000000, 00000000000, 00000000000)
000FFFFB3A0  001801314DB (00000000001, 00000000060, 00000000000, 00000000000)
000FFFFB720  001801575E4 (00000000000, 00000000000, 00000000000, 00000000000)
000FFFFB720  0018012DB7B (7FF85CEF5EC2, 00000004048, 00180249508, 0018013374C)
000FFFFB720  0018012DEAF (003F3D99174, 003F3D990B8, 000000000B6, 0018024948F)
000FFFFB720  001800433E5 (00000000000, 00000000000, 7FF85CF217AA, 00600139370)
000FFFFB860  003F3D87459 (000FFFFB7F0, 0018030BBD0, 000FFFFB870, 006001393D0)
000FFFFB860  00100707C50 (00180142A05, 00000000003, 00100000001, 006000F4780)
000FFFFB860  00100707C2E (00600058A30, 000FFFFB990, 000FFFFB900, 000FFFFC330)
000FFFFB860  001006C025A (000FFFFB920, 00000000000, 00000000000, 00180142A8F)
000FFFFB860  001006B75FA (000FFFFB870, 00000000001, 000FFFFB900, 000FFFFB860)
000FFFFB860  003F3D86EC1 (00000000002, 001FFFFFFFF, 000FFFFBC80, 00600058D50)
00000000000  003F3D951FF (00180154381, 000FFFFC338, 00180157335, 000FFFFBD70)
00000000000  003F3D87648 (000FFFFC330, 00000000000, 006000589B8, 00000000001)
000FFFFC380  001006B40D2 (000FFFFBF88, 000000004B4, 000000517AC, 000FFFFBEE1)
End of stack trace (more stack frames may be present)
Succeeded test: test cases/cmake/5 object library
Failed test during build: 'test cases/cmake/6 object library no dep'
```

**Comment by mensinda** [comment]
Not sure what is going on here. The cygwin test didn't fail with the previous commit (25f16e43f13a4ad9267fb32d5a79fda1957dd2f6) and the last commit basically didn't change anything. Is there a way to rerun the cygwin build?

**Comment by jpakkane** [comment]
I restarted it. Let's see what happens.

**Comment by mensinda** [comment]
The error still occurs and it seems to fail at the same point. This is the code where it is failing for some reason:

```python
        # Wait for CMake to finish
        try:
            self.proc.wait(timeout=2)
        except TimeoutExpired:
            self.proc.terminate()
```

I am going to debug this tomorrow.

**Comment by jpakkane** [comment]
The error message says something about the function `uv_close`. Could this be a Cygwin dependency problem? Some dependency has updated to a broken version or somehow is causing CMake to fail?

**Comment by mensinda** [comment]
Maybe. My current theory is that either CMake doesn't close in this one case for some reason, triggering the timeout and then the terminate fails for some reason...

Or there is a race condition and CMake closes to fast, which causes the wait to throw (which should not happen as far as I know) and then the terminate also throws because the process is already dead. It says `ProcessLookupError: [Errno 3] No such process` in the log for the terminate call so I am assuming that CMake is dead at this point.

I checked the cygwin dependencies and couldn't find any differences. I am not sure if we can just ignore this (although I want to).

**Comment by mensinda** [comment]
OK, I think I know what is going on here:

First, the stack trace with the `uv_close()` assertion is most likely triggered [inside CMake during the server shutdown procedure](https://gitlab.kitware.com/cmake/cmake/blob/master/Source/cmServerConnection.cxx#L71). This most likely happens in every test run because there are a total of 7 assertion errors (6 succeeded tests + 1 failed, and the others were skipped before CMake ran).

Then, there is probably some sort of race condition in the failing test, causing the wait condition to fail, which then triggers the terminate call. However, since CMake has long crashed this function also raises an exception, causing meson to terminate.

I don't think we can do anything about this because CMake itself crashes. However, this isn't really a problem since the crash occurs during the shutdown process and we already have everything we need. I am going to ignore all `terminate()` exceptions for now.

**Comment by jpakkane** [comment]
:crossed_swords: conflicts, but other than that looks good.

**Comment by mensinda** [comment]
Fixed. Let's hope for the CI.

**Comment by mensinda** [comment]
Thanks!

**Pull request merged**